servatrice/src/servatrice_database_interface.cpp

```diff
@@ -132,14 +132,27 @@ bool Servatrice_DatabaseInterface::usernameIsValid(const QString &user, QString
     bool allowNumerics = settingsCache->value("users/allownumerics", true).toBool();
     bool allowPunctuationPrefix = settingsCache->value("users/allowpunctuationprefix", false).toBool();
     QString allowedPunctuation = settingsCache->value("users/allowedpunctuation", "_").toString();
-    error = QString("%1|%2|%3|%4|%5|%6|%7").arg(minNameLength).arg(maxNameLength).arg(allowLowercase).arg(allowUppercase).arg(allowNumerics).arg(allowPunctuationPrefix).arg(allowedPunctuation);
+    QStringList disallowedWords = settingsCache->value("users/disallowedwords", "").toString().split(",", QString::SkipEmptyParts);
+    disallowedWords.removeDuplicates();
+    QStringList disallowedRegExp = settingsCache->value("users/disallowedregexp", "").toString().split(",", QString::SkipEmptyParts);
+    disallowedRegExp.removeDuplicates();
+
+    error = QString("%1|%2|%3|%4|%5|%6|%7|%8|%9").arg(minNameLength).arg(maxNameLength).arg(allowLowercase).arg(allowUppercase).arg(allowNumerics).arg(allowPunctuationPrefix).arg(allowedPunctuation).arg(disallowedWords.join(",")).arg(disallowedRegExp.join(","));
 
     if (user.length() < minNameLength || user.length() > maxNameLength)
         return false;
 
     if (!allowPunctuationPrefix && allowedPunctuation.contains(user.at(0)))
         return false;
 
+    for (const QString &word : disallowedWords) {
+        if (user.contains(word, Qt::CaseInsensitive)) return false;
+    }
+
+    for (const QString &regExp : disallowedRegExp) {
+        if (QRegExp(regExp).exactMatch(user)) return false;
```

Comment:
If we were going to keep the regex support in, we should store compiled regex objects rather than strings. You'd also want to log invalid-regex failures. It's a moot point though since we should take it out